File: python/euler4.py
# A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

# Find the largest palindrome made from the product of two 3-digit numbers.

from decorator import time_and_memory_decorator as tamd
from euler import is_palindrome
import inspect

filename = f'{inspect.getmodule(inspect.stack()[0][0]).__file__[36:-3]}'





@tamd
def find_answer():
    products_of_three = []
    for x in range(100,1000):
        for y in range(x + 1,1000):
            if is_palindrome(x * y):
                products_of_three.append(x * y)
                
    products_of_three.sort()
    return products_of_three[-1]


if __name__ == '__main__':            
    print(filename, ': ', end='')
    answer, mem, time = find_answer()

    print(answer)
    print(mem)
    print(time)

File: cpp/euler4.cpp
// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.

#include "timer.h"
#include "euler.h"
#include <iostream>
#include <ctime>


int main(int argc, char** args)
{
  Timer timer;
  timer.start();

  int *productsOfThree = new int [810000];
  int index = 0;
  for(int i = 100; i < 1000; ++i)
    {
      for(int j = i; j < 1000; ++j)
	{
	  if(isPalindrome(i * j))
	    {
	      productsOfThree[index++] = i * j;
	    }
	}
    }

  int largest = 0;
  for(int i = 0; i < 810000; ++i)
    {
      if(productsOfThree[i] > largest)
	{
	  largest = productsOfThree[i];
	}
      if(productsOfThree[i] == 0)
	{
	  std::cout << largest << std::endl;
	  timer.stop();
	  break;
	}
    }
  
  return 0;
}

